File: app/src/main/java/com/zzzmode/appopsx/ui/permission/AlertInstalledPremActivity.java
package com.zzzmode.appopsx.ui.permission;

import android.content.DialogInterface;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.app.AlertDialog;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewTreeObserver;
import android.view.Window;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;
import com.zzzmode.appopsx.R;
import com.zzzmode.appopsx.ui.BaseActivity;
import com.zzzmode.appopsx.ui.core.LocalImageLoader;
import com.zzzmode.appopsx.ui.model.AppInfo;
import com.zzzmode.appopsx.ui.model.OpEntryInfo;
import com.zzzmode.appopsx.ui.widget.CommonDivderDecorator;
import java.util.List;

/**
 * Created by zl on 2017/5/1.
 */

public class AlertInstalledPremActivity extends BaseActivity implements IPermView {

  public static final String EXTRA_APP = "extra.app";
  private AppInfo appInfo;

  @Override
  protected void onCreate(@Nullable Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);

    appInfo = getIntent().getParcelableExtra(EXTRA_APP);
    if (appInfo == null) {
      finish();
      return;
    }

    initView();

    mPresenter = new PermPresenter(this, appInfo, getApplicationContext());
    mPresenter.setAutoDisabled(false);
    mPresenter.setSortByMode(true);
    mPresenter.setUp();
  }

  private ProgressBar mProgressBar;
  private TextView tvError;
  private PermPresenter mPresenter;
  private AppPermissionAdapter adapter;

  private void initView() {
    final View inflate = getLayoutInflater().inflate(R.layout.activity_opsx, null);

    inflate.findViewById(R.id.appBar).setVisibility(View.GONE);

    tvError = (TextView) inflate.findViewById(R.id.tv_error);
    mProgressBar = (ProgressBar) inflate.findViewById(R.id.progressBar);

    RecyclerView recyclerView = (RecyclerView) inflate.findViewById(R.id.recyclerView);
    recyclerView.setLayoutManager(new LinearLayoutManager(getApplicationContext()));
    recyclerView.addItemDecoration(new CommonDivderDecorator(getApplicationContext()));

    adapter = new AppPermissionAdapter();
    recyclerView.setAdapter(adapter);

    adapter.setListener(new AppPermissionAdapter.OnSwitchItemClickListener() {
      @Override
      public void onSwitch(OpEntryInfo info, boolean v) {

        mPresenter.switchMode(info, v);
      }
    });

    DialogInterface.OnClickListener clickListener = new DialogInterface.OnClickListener() {
      @Override
      public void onClick(DialogInterface dialog, int which) {
        finish();
      }
    };

    AlertDialog alertDialog = new AlertDialog.Builder(this)
        .setCustomTitle(getTitleView())
        .setView(inflate)
        .setPositiveButton(android.R.string.ok, clickListener)
        .create();

    alertDialog.setCanceledOnTouchOutside(false);
    alertDialog.show();
    Window window = alertDialog.getWindow();
    if (window != null) {
      window.setWindowAnimations(0);
    }

    inflate.getViewTreeObserver()
        .addOnGlobalLayoutListener(new ViewTreeObserver.OnGlobalLayoutListener() {
          @Override
          public void onGlobalLayout() {

            int minHeight = getApplicationContext().getResources()
                .getDimensionPixelOffset(R.dimen.dlg_min_height);
            if (inflate.getHeight() < minHeight) {
              ViewGroup.LayoutParams layoutParams = inflate.getLayoutParams();
              layoutParams.height = minHeight;
              inflate.setLayoutParams(layoutParams);
            }
            inflate.getViewTreeObserver().removeOnGlobalLayoutListener(this);

          }
        });

    alertDialog.setOnDismissListener(new DialogInterface.OnDismissListener() {
      @Override
      public void onDismiss(DialogInterface dialog) {
        finish();
      }
    });
  }

  private View getTitleView() {
    View inflate = getLayoutInflater().inflate(R.layout.layout_dlg_title, null);
    ImageView icon = (ImageView) inflate.findViewById(R.id.app_icon);
    TextView tvTitle = (TextView) inflate.findViewById(R.id.tv_title);
    TextView tvInfo = (TextView) inflate.findViewById(R.id.tv_info);

    LocalImageLoader.load(icon, appInfo);
    tvTitle.setText(appInfo.appName);
    tvInfo.setText(appInfo.packageName);

    return inflate;
  }

  @Override
  protected void onDestroy() {
    super.onDestroy();
    if(mPresenter != null) {
      mPresenter.destory();
    }
  }

  @Override
  public void showProgress(boolean show) {
    tvError.setVisibility(View.GONE);
    mProgressBar.setVisibility(show ? View.VISIBLE : View.GONE);

  }

  @Override
  public void showError(CharSequence text) {
    mProgressBar.setVisibility(View.GONE);
    tvError.setVisibility(View.VISIBLE);
    tvError.setText(text);
  }

  @Override
  public void showPerms(final List<OpEntryInfo> opEntryInfos) {
    adapter.setShowConfig(false, false, false);
    adapter.setDatas(opEntryInfos);
    adapter.notifyDataSetChanged();
  }

  @Override
  public void updateItem(OpEntryInfo info) {
    info.mode = info.opEntry.getMode();
    adapter.updateItem(info);
  }
}

File: app/src/main/kotlin/com/zzzmode/appopsx/ui/permission/AlertInstalledPremActivity.kt
package com.zzzmode.appopsx.ui.permission

import android.content.DialogInterface
import android.os.Bundle
import android.support.v7.app.AlertDialog
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.View
import android.view.ViewGroup
import android.view.ViewTreeObserver
import android.view.Window
import android.widget.ImageView
import android.widget.ProgressBar
import android.widget.TextView
import com.zzzmode.appopsx.R
import com.zzzmode.appopsx.ui.BaseActivity
import com.zzzmode.appopsx.ui.core.LocalImageLoader
import com.zzzmode.appopsx.ui.model.AppInfo
import com.zzzmode.appopsx.ui.model.OpEntryInfo
import com.zzzmode.appopsx.ui.widget.CommonDivderDecorator
import kotlinx.android.synthetic.main.activity_alert_opsx.*

/**
 * Created by zl on 2017/5/1.
 */

class AlertInstalledPremActivity : BaseActivity(), IPermView {
    private var appInfo: AppInfo? = null

    private var mProgressBar: ProgressBar? = null
    private var tvError: TextView? = null

    private lateinit var mPresenter: PermPresenter
    private lateinit var adapter: AppPermissionAdapter

    private val titleView: View
        get() {
            val inflate = layoutInflater.inflate(R.layout.layout_dlg_title, null)
            val icon = inflate.findViewById<View>(R.id.app_icon) as ImageView
            val tvTitle = inflate.findViewById<View>(R.id.tv_title) as TextView
            val tvInfo = inflate.findViewById<View>(R.id.tv_info) as TextView

            appInfo?.apply {
                LocalImageLoader.load(icon, this)
                tvTitle.text = appName
                tvInfo.text = "$packageName ($versionName)"
            }


            return inflate
        }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        appInfo = intent?.getParcelableExtra(EXTRA_APP)
        if (appInfo == null) {
            finish()
            return
        }

        initView()

        mPresenter = PermPresenter(this, appInfo!!, applicationContext)
        mPresenter.setAutoDisabled(false)
        mPresenter.setSortByMode(true)
        mPresenter.setUp()
    }

    private fun initView() {
        val inflate = layoutInflater.inflate(R.layout.activity_alert_opsx,null)
        tvError = inflate.findViewById<View>(R.id.tv_error) as TextView
        mProgressBar = inflate.findViewById<View>(R.id.progressBar) as ProgressBar

        val recyclerView = inflate.findViewById<View>(R.id.recyclerView) as RecyclerView
        recyclerView.layoutManager = LinearLayoutManager(applicationContext)
        recyclerView.addItemDecoration(CommonDivderDecorator(applicationContext))

        adapter = AppPermissionAdapter()
        recyclerView.adapter = adapter

        adapter.setListener(object : AppPermissionAdapter.OnSwitchItemClickListener {
            override fun onSwitch(info: OpEntryInfo, v: Boolean) {

                mPresenter.switchMode(info, v)
            }
        })

        val clickListener = DialogInterface.OnClickListener { _, _ -> finish() }

        val alertDialog = AlertDialog.Builder(this)
                .setCustomTitle(titleView)
                .setView(inflate)
                .setPositiveButton(android.R.string.ok, clickListener)
                .create()

        alertDialog.setCanceledOnTouchOutside(false)
        alertDialog.show()
        val window = alertDialog.window
        window?.setWindowAnimations(0)

        inflate.viewTreeObserver
                .addOnGlobalLayoutListener(object : ViewTreeObserver.OnGlobalLayoutListener {
                    override fun onGlobalLayout() {

                        val minHeight = applicationContext.resources
                                .getDimensionPixelOffset(R.dimen.dlg_min_height)
                        if (inflate.height < minHeight) {
                            val layoutParams = inflate.layoutParams
                            layoutParams.height = minHeight
                            inflate.layoutParams = layoutParams
                        }
                        inflate.viewTreeObserver.removeOnGlobalLayoutListener(this)

                    }
                })

        alertDialog.setOnDismissListener { finish() }
    }

    override fun onDestroy() {
        super.onDestroy()
        mPresenter.destory()
    }

    override fun showProgress(show: Boolean) {
        tvError?.visibility = View.GONE
        mProgressBar?.visibility = if (show) View.VISIBLE else View.GONE

    }

    override fun showError(text: CharSequence) {
        mProgressBar?.visibility = View.GONE
        tvError?.visibility = View.VISIBLE
        tvError?.text = text
    }


    override fun showPerms(opEntryInfos: List<OpEntryInfo>) {
        adapter.setShowConfig(false, false, false)
        adapter.datas = opEntryInfos
        adapter.notifyDataSetChanged()
    }

    override fun updateItem(info: OpEntryInfo) {
        info.mode = info.opEntry.mode
        adapter.updateItem(info)
    }

    companion object {

        const val EXTRA_APP = "extra.app"
    }
}
